商品出品機能

miyazato1872 opened the pull request:
# What
コードレビューと添付URLの動画と画像の確認をお願いします。

 ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/783911ef0ff141dcf34573f67d45cc2e

 価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/7996e45fa4c22fee45df73bcf60664e6

 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/0bfae67cf7f201799bd49df1c58f00c5

 入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/186731c9779e1495db2ff9aed49071d6

 ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/9f8977d7ef5d1f25774ee0626daa71ab

 テスト結果の画像
https://gyazo.com/f55e204990c0ac39473ea6226b177aa4

# Why
商品出品機能の実装のためです。

miyazato1872 committed:
Active Storageと画像加工ツールの導入

miyazato1872 committed:
Itemモデルとテーブル、カラムを作成

アソシエーションもOK

miyazato1872 committed:
データベース、newアクション、newビューファイル作成

miyazato1872 committed:
ActiveHash導入とビューファイル修正

miyazato1872 committed:
new.html.erb内のプルダウンを作成

miyazato1872 committed:
itemテーブルへの保存機能とエラーメッセージ表示

miyazato1872 committed:
画像が空なら保存できないバリデーション設定

miyazato1872 committed:
ログイン状態の有無によって、商品出品ページへの遷移可否を設定

miyazato1872 committed:
JavaScriptによる販売手数料と利益の自動表示の実装

miyazato1872 committed:
Itemモデルのテストコード

miyazato1872 committed:
Rubocopを実行

miyazato1872 committed:
index.htmlのビューファイルの出品ボタンを修正

miyazato1872 commented:
自分宛
届いているか確認のためのコメントです

wad-sho commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,26 @@
+
+  def item_params
+    params.require(:item).permit(:image, :name, :description, :category_id, :status_id, :shipping_charge_id, :area_id,
+                                 :days_to_ship_id, :price, :user_id).merge(user_id: current_user.id)
```
` :user_id
`
permit内のuser_idは削除いたしましょう。
user_idはフォームから送られるパラメーターではないためです。

wad-sho commented on app/models/item.rb:
```diff
@@ -1,2 +1,30 @@
+  validates :status_id,           presence: true
+  validates :shipping_charge_id,  presence: true
+  validates :area_id,             presence: true
+  validates :days_to_ship_id,     presence: true
```
```
  validates :category_id,         presence: true
  validates :status_id,           presence: true
  validates :shipping_charge_id,  presence: true
  validates :area_id,             presence: true
  validates :days_to_ship_id,     presence: true
```

Activehashに対してpresence: trueは今回の実装では不要となりますので削除いたしましょう！

wad-sho commented on app/models/item.rb:
```diff
@@ -1,2 +1,30 @@
+  validates :days_to_ship_id,    numericality: { other_than: 1, message: "can't be blank" }
+
+  validates_inclusion_of :price, in: 300..9_999_999, message: 'Price is invalid', allow_blank: true
+  validates :price, format: { with: /\A[0-9]+\z/ }, allow_blank: true
```
validates :price, format: { with: /\A[0-9]+\z/ }, allow_blank: true

integer型のカラムに対して正規表現は無効となるため、上記バリデーションは取り除いておきましょう。

wad-sho commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,118 @@
+        @item.price = '１２３あいう'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price Price is invalid')
+      end
```
userが紐づいていなければ登録できないテストを追記いたしましょう！

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
コメントいたしましたので、ご確認頂けますと幸いでございます。

ご修正後再度レビュー依頼のほどよろしくお願いいたします！

miyazato1872 committed:
再コードレビュー1回目

・permit内のuser_idは削除
・Activehashに対してpresence: trueは今回の実装では不要とるため削除
・integer型のカラムに対して正規表現は無効となるため、対象のバリデーションを削除
・userが紐づいていなければ登録できないテストを追記

miyazato1872 commented:
# what
・permit内のuser_idは削除
・Activehashに対してpresence: trueは今回の実装では不要とるため削除
・integer型のカラムに対して正規表現は無効となるため、対象のバリデーションを削除
・userが紐づいていなければ登録できないテストを追記

# Why
コードレビュー再提出のため

sh10n7 reviewed (commented):
レビュー依頼ありがとうございます。
今回バリデーションの修正やテストコードの追記対応をいただいておりますので、挙動確認のために**最新のテスト画像**を添付いただけますでしょうか？

上記こちらのプルリクエストへ添付いただいた後、再度レビュー依頼フォームよりご依頼をお願いいたします。

ご対応のほど、よろしくお願いいたします。

miyazato1872 commented:
# 最新のテスト画像を添付いたします。
https://gyazo.com/2e494d42473720a53a460fb02d7c83d6

よろしくお願いいたします。

zonodance commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,124 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Days to ship can't be blank")
+      end
+
```
＜修正任意＞
・価格に対するエラーメッセージに関して
現状、"Price Price is invalid"のように、主語であるPriceが2つ並んでいる状況かと存じます。
こちらはitem.rbを修正することで、適切な文面になりますので、ぜひチャレンジしていただけますと幸いです。

zonodance reviewed (commented):
商品出品機能のレビュー依頼ありがとうございました。
こちら`LGTM`とさせて頂きます。

任意での修正をコメントしております。
こちら`再度レビュー依頼は出さなくて大丈夫です`ので、修正後、コミットとプッシュして頂けましたら、
マージ後、ローカルのmasterブランチに忘れずにプルし、次の実装に取り組んで頂ければと思います。

今後の実装も応援しております！

miyazato1872 committed:
価格に対するエラーメッセージの修正

miyazato1872 commented:
LGTMありがとうございます！
修正に関しても承知いたしました。
